cSploit/jni/libncp/man/ncpmount.8
.TH NCPMOUNT 8 12/04/1998 ncpmount ncpmount
.SH NAME
ncpmount, mount.ncp, mount.ncpfs \- mount volume(s) from a specified NetWare fileserver.
.SH SYNOPSIS
.B ncpmount
[ \fB-h\fP ] [ \fB-S\fP \fIserver\fP ] [ \fB-U\fP \fIuser name\fP ]
[ \fB-P\fP \fIpassword\fP | \fB-n\fP ] [ \fB-C\fP ] 
[ \fB-c\fP \fIclient name\fP ] [ \fB-u\fP \fIuid\fP ] [ \fB-g\fP \fIgid\fP ] 
[ \fB-f\fP \fIfile mode\fP ] [ \fB-d\fP \fIdir mode\fP ]
[ \fB-V\fP \fIvolume\fP ] [ \fB-t\fP \fItime_out\fP ]
[ \fB-r\fP \fIretry_count\fP ] [ \fB-b\fP ] [ \fB-i\fP \fIlevel\fP ]
[ \fB-v\fP ] [ \fB-m\fP ] [ \fB-y\fP \fIiocharset\fP ]
[ \fB-p\fP \fIcodepage\fP ] [ \fB-N\fP \fIignored namespace\fP ]
[ \fB-2\fP | \fB-3\fP | \fB-4\fP ] [ \fB-s\fP ] [ \fB-A\fP \fIdns name\fP ]
mount-point

.B mount.ncp
remote-server-and-user
mount-point
[ \fB-n\fP ] [ \fB-v\fP ] [ \fB-o\fP \fImount_options\fP ]


.SH DESCRIPTION
This program is used to mount volumes of the specified NetWare Fileserver
under the specified mount point.

.B ncpfs
is a linux filesystem which understands the NCP protocol. This is the
protocol Novell NetWare clients use to talk to NetWare servers. ncpfs
was inspired by \fBlwared\fP, a free NetWare emulator for Linux written
by Ales Dryak. See ftp://klokan.sh.cvut.cz/pub/linux for this very 
interesting program.

\fBncpmount\fP, when invoked with all appropriate arguments, attaches and
logs into specified server and mounts all volumes (or one volume or subtree)
from server under the specified mount point.
.B ncpmount
when invoked without any arguments specifying the fileserver, user id and
password checks the file
.I $HOME/.nwclient
to find a file server, a user name and possibly a password to use for the
specified mount point. See nwclient(5) for more information. Please note
that the access permissions of .nwclient MUST be 600, for security reasons.

.SH OPTIONS
.B mount-point
.RS 3
.B mount-point
is the directory you want to mount the filesystem over. Its function is the
the same as for a normal mount command.

If the real uid of the caller is not root,
.B ncpmount
checks whether the user is allowed to mount a filesystem on the
mount-point. So it should be safe to make 
.B ncpmount
setuid root. The filesystem stores the uid of the user who called
ncpmount. So
.B ncpumount
can check whether the caller is allowed to unmount the filesystem.
.RE

.B -S
.I server
(mount option \fBserver=\fP or part before
.B /
in \fIremote-server-and-user\fP)
.RS 3
.B server
is the name of the server you want to use.
.RE

.B -h
.RS 3
.B -h
is used to print out a short help text.
.RE

.B -C
(mount option \fBnoupcasepasswd\fP)
.RS 3
By default passwords are converted to uppercase before they are sent
to the server because most servers require this. This option disables this
feature ensuring that passwords are sent without any case conversion.
.RE

.B -n
(mount option \fBnopasswd\fP)
.RS 3
.B -n
must be specified for logins that do not have a password configured.
This option means do not update /etc/mtab if there is option
.B -o
on command line. You must use
.B -o nopasswd
in this case.
.RE

.B passwdfile=file
(only mount option)
.RS 3
If you want specify password and you do not want store it into world readable
/etc/fstab, you can use this option.
.I file
then contains lines in form SERVER/USER:PASSWORD:other_data (other_data are
currently unused)
.RE

.B pass-fd=fd
(only mount option)
.RS 3
If you want to pass password in secure way to ncpmount, you can pass it through
specified \fIfd\fP.
.RE

.B -P
.I password
(mount option \fBpasswd=\fP)
.RS 3
specifies the password to use for the Netware user id.

If neither
.B -n
nor the
.B -P
nor the
.B passwdfile=
nor the
.B pass-fd=
arguments are specified ncpmount will prompt for a password. This
makes it difficult to use in scripts such as /etc/rc. If you want to
have ncpmount work automatically from a script you must include the
appropriate option and be very careful to ensure that appopriate file
permissions are set for the script that includes your password to
ensure that others can not read it.
.RE

.B -U
.I user name
(mount option \fBuser=\fP or rest of \fIremote-server-and-user\fP after 
\fB/\fP)
.RS 3
Specifies the Netware user id to use when logging in to the fileserver. If
this option is not specified then ncpmount will attempt to login to the
fileserver using the Linux login id of the user invoking ncpmount.
.RE

.B -m
(mount option \fBmultiple\fP)
.RS 3
Normally,
.B ncpmount
limits number of connections from client to server to one per unique user
name. If you want mount more than one connection with same username and
server, you must specify \fB-m\fP.
.RE

.B -u
.I uid,
.B -g
.I gid
(mount option \fBuid=\fP and \fBgid=\fP)
.RS 3
ncpmount does not yet implement a scheme for mapping NetWare users/groups
to Linux users/groups. Linux requires that each file has an owner and group id.
With
.B -u
and
.B -g
you can tell ncpmount which id's it should assign to the files in the
mounted directory.

The defaults for these values are the current uid and gid.
.RE

.B -c
.I user name
(mount option \fBowner=\fP)
.RS 3
.B -c
names the user who is the 
.I owner
of the connection, where owner does not refer to file ownership (that
"owner" is set by the -u argument), but the owner of the mount, ie: who
is allowed to call ncpumount on this mount. The default owner of the
connection and the mount is the user who called ncpmount. This option
allows you to specify that some other user should be set as the owner.

In this this way it is possible to mount a public read-only directory,
but to allow the lp daemon to print on NetWare queues. This is
possible because only users who have write permissions on a directory
may issue ncp requests over a connection. The exception to this rule
is the 'mount owner', who is also granted 'request permission'.
.RE

.B -f
.I file mode,
.B -d
.I dir mode
(mount option \fBmode=\fP (or \fBfilemode=\fP) and \fBdirmode=\fP)
.RS 3
Like 
.B -u
and \fB-g\fP,
these options are used to determine what permissions should be assigned
files and directories of the mounted volumes. The values must be specified
as octal numbers. The default values are taken from the current umask, where
the file mode is the current umask, and the dir mode adds execute permissions
where the file mode gives read permissions.

Note that these permissions can differ from the rights the server
gives to us. If you do not have write permissions on the server, you
can very well choose a file mode that tells that you have. This
certainly cannot override the restrictions imposed by the server.
.RE

.B -V
.I volume
(mount option \fBvolume=\fP)
.RS 3
There are 2 general ways you can mount a NetWare server's disk space:
Either you can mount all volumes under one directory, or you can mount
only a single volume.

When you choose to mount the complete disk space at once, you have the
advantage that only one Linux mount point and only one
NetWare connection is used for all the volumes of this server. Both of
these are limited resources. (Although raising the number of Linux
mount points is significantly cheaper than raising the number of
available NetWare connections ;-))

When you specify to mount a single volume by using the option
.B -V
.I volume,
you have the big advantage that nfsd is able to re-export this mounted
directory. You must invoke 
.B nfsd
and
.B mountd
with the option
.I --re-export
to make nfsd re-export ncpfs mounted directories. This uses one Linux
mount point and one NetWare connection per mounted volume. Maybe
sometime in the future I will make it possible to mount all volumes on
different mount points, using only one connection.
.RE

.B -t
.I time_out
(mount option \fBtimeo=\fP or \fBtimeout=\fP)
.RS 3
With 
.B -t
you can adjust the time ncpfs waits for the server to answer a request
it sent. Use the option to raise the timeout value when your ncpfs
connections seem to be unstable although your servers are well
up. This can happen when you have very busy servers, or servers that
are very far away.

.I time_out
is specified in 1/100s, the current default value is 60.
.RE

.B -r
.I retry_count
(mount option \fBretry=\fP)
.RS 3
As \fB-t\fP, \fB-r\fP can be used to tune the ncpfs connection to the
server. With \fIretry_count\fP you can specify how many times ncpfs 
will attempt to send a packet to the server before it decides the 
connection is dead. The current default value is 5.

Currently ncpfs is not too clever when trying to find out that
connections are dead. If anybody knows how to do that correctly, as it
is done by commercial workstations, please tell me.
.RE

.B -y
.I iocharset
(mount option \fBiocharset=\fP)
.RS 3
You can specify character translation rules for converting names from
unicode to your desktop (it works together with \fB-p\fP).
.I iocharset
is charset name, for example \fIiso-8859-1\fP.
.RE

.B -p
.I codepage
(mount option \fBcodepage=\fP)
.RS 3
You can specify character translation rules for converting names from
Netware encoding to unicode (it works together with \fB-y\fP).
.I codepage
is codepage name, for example \fIcp437\fP.
.RE

\fB-b\fP (mount option \fBbindery\fP)
.RS 3
If you are connecting to NetWare 4 or NetWare 5 through bindery emulation
instead of NDS, you must specify this option.
.RE

.B -i
.I level
(mount option \fBsignature=\fP\fIlevel\fP)
.RS 3
Enables packet signing. \fIlevel\fP is from 0 to 3: 0 means disable, 1 means
sign if server needs it, 2 means sign if server allows it and 3 means sign
packets always.
.RE

.B -v
.RS 3
Print ncpfs version number. It has another meaning (\fIverbose\fP) if you specify
.B -o
on command line. If you are interested in version, type
.B ncpmount -v
without another options.
.RE

\fB-A\fP \fIdns name\fP (mount option \fBipserver=\fP\fIdns name\fP)
.RS 3
When you are mounting volumes from NetWare 5 server over UDP, you must
specify \fIdns name\fP of server here and \fIlogical server name\fP in
\fB-S\fP (or in \fBserver=\fP). This name is used to switch \fBncpmount\fP
into UDP mode and to specify server to connect. Currently, \fBDNS\fP is
only supported \fBIP\fP name resolution protocol. There is currently no
support for \fBSLP\fP.
.RE

\fB-N\fP \fIignored namespace\fP (mount option \fBnonfs\fP and \fBnolong\fP)
.RS 3
\fBncpfs\fP supports \fINFS\fP, \fILONG\fP (\fIOS/2\fP) and \fIDOS\fP namespace
on NetWare volumes. If you do not want to use \fINFS\fP or \fILONG\fP namespace
(because of bugs in (server) code or for backward compatibility), you must
specify these ignored namespaces in mount parameters.
.RE

\fB-2\fP
.RS 3
If you have unusual \fBncpfs\fP code in kernel and \fBncpmount\fP is not
able to autodetect it, use this option. It switches \fBncpmount\fP to 
\fBncpfs\fP interface version 2. This interface was used in 2.0.x kernels,
does not support NCP/UDP, does not have NDS authentication info storage
and uses only 16bit uid/gid.
.RE

\fB-3\fP
.RS 3
If you have unusual \fBncpfs\fP code in kernel and \fBncpmount\fP is not

cSploit/jni/libncp/sutil/ncpmount.c
/*
    ncpmount.c
    Copyright (C) 1995, 1997 by Volker Lendecke

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

    Revision history:

	0.00  1995			Volker Lendecke
		Initial release.

	0.01  1996, January 20		Steven N. Hirsch <hirsch@emba.uvm.edu>
		If the ncpfs support is not loaded and we are using kerneld to
		autoload modules, then we don't want to do it here.  I added
		a conditional which leaves out the test and load code.

		Even if we _do_ want ncpmount to load the module, passing a
		fully-qualified pathname to modprobe causes it to bypass a 
		path search.  This may lead to ncpfs.o not being found on
 		some systems.

	0.02  1998, November 28		Wolfram Pienkoss <wp@bszh.de>
		NLS added. Thanks Petr Vandrovec for the idea of charset select
		and many hints.

	0.03  1999, September 9		Petr Vandrovec <vandrove@vc.cvut.cz>
 		Added ttl option.
		Reverted back to exec("/sbin/nwmsg") for incomming messages. 
		This allows tknwmsg work again...

	0.04  1999, September 9		Paul Rensing <paulr@dragonsys.com>
		Added resetting behavior for SIGTERM. This fixes ncpmount from 
		perl scripts.

	1.00  1999, November 20		Petr Vandrovec <vandrove@vc.cvut.cz>
		Added license.
		Replaced nds_get_tree_name with NWIsDSServer.

	1.01  2000, January 16		Petr Vandrovec <vandrove@vc.cvut.cz>
		Changes for 32bit uids.
		
	1.02  2000, April 4		Petr Vandrovec <vandrove@vc.cvut.cz>
		Changed /sbin/nwmsg to $(sbindir)/nwmsg (recomended by
			MandrakeSoft).

	1.03  2000, April 28		Petr Vandrovec <vandrove@vc.cvut.cz>
		Silently ignore 'noauto' option, mount-2.10f (at least)
			passes it down to mount.ncp.

	1.04  2000, June 25		Petr Vandrovec <vandrove@vc.cvut.cz>
		Add NCP/TCP support.

	1.05  2000, October 28		Petr Vandrovec <vandrove@vc.cvut.cz>
		Added nfsextras support.
		Changed some constants to symbolic names (SA_*)

	1.06  2001, January 7		Patrick Pollet <patrick.pollet@insa-lyon.fr>
		Made all error exit codes differents to help debug Pam module

	1.07  2001, February 25		Petr Vandrovec <vandrove@vc.cvut.cz>
		Added UNIX socket handling for global features.

	1.08  2001, September 23	Petr Vandrovec <vandrove@vc.cvut.cz>
		Fix multiple mounts when using UDP or TCP.

	1.09  2002, January 20		Petr Vandrovec <vandrove@vc.cvut.cz>
		Verify uid/gid/owner arguments carefully. It should not be
		security problem, but who knows which program derives some 
		security information from file uid/gid...

	1.10  2002, February 27		Petr Vandrovec <vandrove@vc.cvut.cz>
		Erase password quickly from command line.
		Clear command line completely.
		Simpler command line handling.

	1.11  2002, April 19		Peter Schwindt <schwindt@ba-loerrach.de>
		Add option -A to the help text.

 */

#define NCP_OBSOLETE

#include "ncpm_common.h"
#include "mount_login.h"

#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <ncp/ext/socket.h>
#include <sys/param.h>
#include <netinet/in.h>
#include <netdb.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/errno.h>
#include <sys/time.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <ctype.h>
#include <stdlib.h>
#include <stdarg.h>
#include <sys/mount.h>
#include <mntent.h>
#include <ncp/kernel/ipx.h>
#include <ncp/nwclient.h>
#include <sys/ioctl.h>
#ifdef CONFIG_NATIVE_UNIX
#include <linux/un.h>
#endif

#include "private/libintl.h"
#define _(X) gettext(X)

#include "../lib/ncpi.h"

static void usage(void);
static void help(void);

void veprintf(const char* message, va_list ap) {
	vfprintf(stderr, message, ap);
}

static void opt_set_tree(struct ncp_mount_info* info, const char* param) {
	if (strlen(param) > MAX_TREE_NAME_CHARS) {
		errexit(65, _("Specified tree name `%s' is too long\n"), param);
	}
	info->tree = param;
}

static void opt_set_bindery(struct ncp_mount_info* info, unsigned int param) {
	info->force_bindery_login = param;
}

static void opt_set_volume(struct ncp_mount_info* info, const char* param) {
	info->volume = param;
}

static void opt_set_allow_multiple(struct ncp_mount_info* info, unsigned int param) {
	info->allow_multiple_connections = param;
}

static void opt_set_auth(struct ncp_mount_info* info, const char* param) {
	info->auth_src = param;
}

static struct optinfo opts[] = {
	{'T', "tree",		FN_STRING,	opt_set_tree,		0},
	{'b', "bindery",	FN_NOPARAM,	opt_set_bindery,	1},
	{'V', "volume",		FN_STRING,	opt_set_volume,		0},
	{'m', "multiple",	FN_NOPARAM,	opt_set_allow_multiple,	1},
	{'a', "auth",		FN_STRING,	opt_set_auth,		0},
	{0,   NULL,		0,		NULL,			0}
};

int
main(int argc, char *argv[])
{
	struct ncp_mount_info info;
	struct stat st;
	char *mount_name;
#ifdef NDS_SUPPORT
	NWDSContextHandle ctx;
	NWCONN_HANDLE auth_conn;
#endif

	int result;

	long err;

	struct ncp_conn_spec spec;

	struct ncp_conn *conn;

	int opt;

#ifdef CONFIG_NATIVE_UNIX
	struct sockaddr_un server_un;
#endif

	char *tmp_mount;

	int mount_ncp = 0;
	int opt_n = 0;
	int opt_v = 0;

	verify_argv(argc, argv);

	setlocale(LC_ALL, "");
	bindtextdomain(NCPFS_PACKAGE, LOCALEDIR);
	textdomain(NCPFS_PACKAGE);
	
	init_mount_info(&info);

	while ((opt = getopt(argc, argv, "a:CS:U:c:u:g:f:d:P:nh?vV:t:r:o:"
		"sN:y:p:bi:mA:T:"
#ifdef MOUNT2
		"2"
#endif
#ifdef MOUNT3
		"345"
#endif
	        )) != EOF)
	{
		char tmpopt[2];

		switch (opt)
		{
		case 'C':
		case 's':
		case 'b':
		case 'm':
			optarg = NULL;
		case 'a':
		case 'S':
		case 'T':
		case 'U':
		case 'c':
		case 'u':
		case 'g':
		case 'f':
		case 'd':
		case 'V':
		case 't':
		case 'r':
		case 'i':
		case 'A':
		case 'y':
		case 'p':
			tmpopt[0] = opt;
			tmpopt[1] = 0;
			proc_option(opts, &info, tmpopt, optarg);
			break;
		case 'P':
			if (!mount_ncp) opt_n = 0;	/* clear -n (nopasswd) */
			tmpopt[0] = opt;
			tmpopt[1] = 0;
			proc_option(opts, &info, tmpopt, optarg);
			break;
		case 'n':
			opt_n = 1;	/* no passwd or no update /etc/fstab */
			break;
		case 'h':
		case '?':
			help();
			exit(27);
		case 'v':
			opt_v = 1;	/* version or verbose */
			break;	
#ifdef MOUNT2
		case '2':
			info.version = 2;
			break;
#endif /* MOUNT2 */
#ifdef MOUNT3
		case '3':
			info.version = 3;
			break;
		case '4':
			info.version = 4;
			break;
		case '5':
			info.version = 5;
			break;
#endif /* MOUNT3 */
		case 'N':
			{
				char *inp;
				char *out;

				inp = optarg;

				while ((out = strtok(inp, ",;:"))!=NULL) {
					inp=NULL;
					if (!strcasecmp(out, "OS2"))
						proc_option(opts, &info, "noos2", NULL);
					else if (!strcasecmp(out, "LONG"))
						proc_option(opts, &info, "nolong", NULL);
					else if (!strcasecmp(out, "NFS"))
						proc_option(opts, &info, "nonfs", NULL);
					else {
						errexit(128, _("Unknown namespace \"%s\"\n"), out);
					}
				}
			};
			break;
		case 'o':
			{
				char* arg;

				arg = optarg;
				mount_ncp = 1;
				while (arg) {
					char* end;
					char* parm;

					arg += strspn(arg, " \t");
					if (!*arg) break;
					end = strchr(arg, ',');
					if (end) *end++ = 0;

					parm = strchr(arg, '=');
					if (parm) *parm++ = 0;

					proc_option(opts, &info, arg, parm);
					arg = end;
				};
			}
			break;
		default:
			usage();
			return -1;
		}
	}
	
	if (opt_n && !mount_ncp) {
		opt_n = 0;
		proc_option(opts, &info, "passwd", "");
	}
	if (opt_v && !mount_ncp) {
		fprintf(stderr, _("ncpfs version %s\n"), NCPFS_VERSION);
		exit(30);
	}
	if (optind < argc - 1) {
		char* s;
		char* u;

		s = strdup(argv[optind++]);
		if (s) {
			u = strchr(s, '/');
			if (u) *u++ = 0;
		} else
			u = NULL;
		if (!info.server && !info.tree) info.server = s;
		if (!info.user) info.user = u;
	}
	if (info.server && info.tree) {
		errexit(66, _("Both tree and server name were specified. It is not allowed.\n"));
	}
	if (myeuid != 0)
	{
		errexit(26, _("%s must be installed suid root\n"), progname);
	}
	if (optind != argc - 1)
	{
		usage();
		return -1;
	}

	realpath(argv[optind], mount_point);

	if (chdir(mount_point))
	{
		errexit(31, _("Could not change directory into mount target %s: %s\n"),
			mount_point, strerror(errno));
	}
	if (stat(".", &st) == -1)
	{
		errexit(31, _("Mount point %s does not exist: %s\n"),
			mount_point, strerror(errno));
	}
	if (mount_ok(&st) != 0)
	{
		errexit(32, _("Cannot to mount on %s: %s\n"),
			mount_point, strerror(errno));
	}

	get_passwd(&info);

	if (myuid != info.mdata.mounted_uid) {
		if (setreuid(info.mdata.mounted_uid, -1)) {
			errexit(76, _("Cannot impersonate as requested: %s\n"), strerror(errno));
		}
	}
#ifdef NDS_SUPPORT
	if (info.tree) {
		static char vol[64];
		static char serv[64];
		nuint32 ctxf;
		
		if (!info.volume) {
			errexit(70, _("You must specify NDS volume name if you specified tree name.\n"));
		}
		err = NWDSCreateContextHandle(&ctx);
		if (err) {
			mycom_err(err, _("in create context"));
			exit(68);
		}
		if (!NWDSGetContext(ctx, DCK_FLAGS, &ctxf)) {
			ctxf |= DCV_XLATE_STRINGS | DCV_DEREF_ALIASES | DCV_TYPELESS_NAMES | DCV_DEREF_BASE_CLASS;
			NWDSSetContext(ctx, DCK_FLAGS, &ctxf);
		}
		err = NWCCOpenConnByName(NULL, info.tree, NWCC_NAME_FORMAT_NDS_TREE,
			NWCC_OPEN_NEW_CONN, 
			NWCC_RESERVED, &conn);
		if (err) {
			NWDSFreeContext(ctx);
			mycom_err(err, _("in tree search"));
			exit(67);
		}
		err = nds_login_auth(conn, info.user, info.password);
		if (err) {
			NWCCCloseConn(conn);
			NWDSFreeContext(ctx);
			mycom_err(err, _("in nds login"));
			exit(68);
		}
		NWDSAddConnection(ctx, conn);
		err = NWCXGetNDSVolumeServerAndResourceName(ctx,
			info.volume, serv, sizeof(serv),
			vol, sizeof(vol));
		if (err) {
			NWDSFreeContext(ctx);
			NWCCCloseConn(conn);
			mycom_err(err, _("in volume search"));
			exit(69);
		}
		NWDSFreeContext(ctx);
		NWCCCloseConn(conn);
		info.server = serv;
		strcpy(info.NWpath + 2, vol);
		info.NWpath[0] = 1;
		info.NWpath[1] = strlen(vol);
		info.pathlen = strlen(vol) + 2;
	} else 
#endif
	{
		if (info.volume) {
			int ln;

			ln = ncp_path_to_NW_format(info.volume, info.NWpath, sizeof(info.NWpath));
			info.remote_path = info.volume;
			if (ln < 0) {
				errexit(18, _("Volume path `%s' is invalid: `%s'\n"), info.volume, strerror(-ln));
			};
			info.pathlen = ln;
			if (info.pathlen == 1) {
				info.mdata.mounted_vol = "";
				info.remote_path = "/";
			} else if (info.NWpath[0] != 1) {
				info.mdata.mounted_vol = "dummy";
			} else if (strlen(info.volume) > NCP_VOLNAME_LEN) {
				errexit(19, _("Volume name `%s' is too long\n"), info.volume);
			} else {
				info.mdata.mounted_vol = info.volume;
			}
		}
	}
	err = NWDSCreateContextHandle(&ctx);
	if (err) {
		mycom_err(err, _("in create context"));
		exit(113);
	}
	if (info.auth_src) {
		NWDSChar user[MAX_DN_BYTES];

		err = ncp_open_mount(info.auth_src, &auth_conn);
		if (err) {
			NWDSFreeContext(ctx);
			mycom_err(err, _("opening mount %s"), info.auth_src);
			exit(114);
		}
		NWDSAddConnection(ctx, auth_conn);
		err = NWDSWhoAmI(ctx, user);
		if (err) {
			NWDSFreeContext(ctx);
			NWCCCloseConn(auth_conn);
			mycom_err(err, _("retrieving user name"));
			exit(117);
		}
		memset(&spec, 0, sizeof(spec));
		strncpy(spec.server, info.server, sizeof(spec.server));
		strncpy(spec.user, user, sizeof(spec.user));
	} else {
		if ((err = ncp_find_conn_spec3(info.server, info.user, info.password, 1, 
				info.mdata.uid, 1, &spec))
		    != 0)
		{
			NWDSFreeContext(ctx);
			mycom_err(err, _("in find_conn_spec"));
			exit(33);
		}
		if (info.upcase_password != 0)
		{
			str_upper(spec.password);
		}
		auth_conn = NULL;
	}
	info.mdata.server_name = spec.server;

	if (info.mdata.dir_mode == 0)
	{
		info.mdata.dir_mode = info.mdata.file_mode;
		if ((info.mdata.dir_mode & S_IRUSR) != 0)
			info.mdata.dir_mode |= S_IXUSR;
		if ((info.mdata.dir_mode & S_IRGRP) != 0)
			info.mdata.dir_mode |= S_IXGRP;
		if ((info.mdata.dir_mode & S_IROTH) != 0)
			info.mdata.dir_mode |= S_IXOTH;
	}
#if defined(CONFIG_NATIVE_IP) || defined(CONFIG_NATIVE_UNIX)
	if (info.server_name) {
#if defined(CONFIG_NATIVE_UNIX)
		if (info.server_name[0] == '/') {
			server_un.sun_family = AF_UNIX;
			memcpy(server_un.sun_path, "\000ncpfs", 6);
			memcpy(&info.mdata.serv_addr, &server_un, 16);
		} else
#endif
#if defined(CONFIG_NATIVE_IP)		
		{
			nuint tran;

			if ((!info.allow_multiple_connections)&&
			    ((tmp_mount = ncp_find_permanent(&spec)) != NULL))
			{
				NWDSFreeContext(ctx);
				if (auth_conn) {
					NWCCCloseConn(auth_conn);
				}
				fprintf(stderr,
					_("You already have mounted server %s\nas user "
					"%s\non mount point %s\n"), spec.server, spec.user,
					tmp_mount);
				exit(35);
			}

			tran = NT_UDP;
			if (info.nt.valid && info.nt.value == NT_TCP) {
				tran = info.nt.value;
			}
			err = ncp_find_server_addr(&info.server_name, OT_FILE_SERVER, (struct sockaddr*)&info.mdata.serv_addr,
					sizeof(info.mdata.serv_addr), tran);
			if (err) {
				NWDSFreeContext(ctx);
				if (auth_conn) {
					NWCCCloseConn(auth_conn);
				}
				eprintf(_("Get host address `%s': %s\n"), info.server_name, strnwerror(err));
				exit(1);
			}
		}
#else
		{
			goto ncpipx;
		}
#endif		
	} else 
#endif	/* CONFIG_NATIVE_IP */
	{
#if defined(CONFIG_NATIVE_UNIX) && !defined(CONFIG_NATIVE_IP)
ncpipx:;
#endif
		if ((!info.allow_multiple_connections)&&
		    ((tmp_mount = ncp_find_permanent(&spec)) != NULL))
		{
			NWDSFreeContext(ctx);
			if (auth_conn) {
				NWCCCloseConn(auth_conn);
			}
			fprintf(stderr,
				_("You already have mounted server %s\nas user "
				"%s\non mount point %s\n"), spec.server, spec.user,
				tmp_mount);
			exit(35);
		}
		if ((err = ncp_find_fileserver(spec.server, (struct sockaddr*)&info.mdata.serv_addr, sizeof(info.mdata.serv_addr))) != 0)
		{
			NWDSFreeContext(ctx);
			if (auth_conn) {
				NWCCCloseConn(auth_conn);
			}
			mycom_err(err, _("when trying to find %s"),
				spec.server);
			exit(36);
		}
	}

	err = proc_buildconn(&info);
	if (err) {
		NWDSFreeContext(ctx);
		if (auth_conn) {
			NWCCCloseConn(auth_conn);
		}
		exit(err);
	}

	info.mdata.mount_point = mount_point;

	if (asprintf(&mount_name, "%s/%s", spec.server, spec.user) < 0) {
		NWDSFreeContext(ctx);
		if (auth_conn) {
			NWCCCloseConn(auth_conn);
		}
		errexit(85, _("Cannot allocate memory for mtab entry: %s\n"), strerror(ENOMEM));
	}

	result = ncp_mount(mount_name, &info);
	if (result)
	{
		free(mount_name);
		NWDSFreeContext(ctx);
		if (auth_conn) {
			NWCCCloseConn(auth_conn);
		}
		mycom_err(result, _("in mount(2)"));
		exit(51);
	}
	err = proc_aftermount(&info, &conn);
	if (err) {
		proc_ncpm_umount(info.mdata.mount_point);
		free(mount_name);
		NWDSFreeContext(ctx);
		if (auth_conn) {
			NWCCCloseConn(auth_conn);
		}
		exit(err);
	}
#ifdef NDS_SUPPORT
	if (info.auth_src != NULL) {
		err = NWDSAuthenticateConn(ctx, conn);
		if (err) {
			proc_ncpm_umount(info.mdata.mount_point);
			free(mount_name);
			NWDSFreeContext(ctx);
			if (auth_conn) {
				NWCCCloseConn(auth_conn);
			}
			mycom_err(err, _("in authenticate connection"));
			exit(115);
		}
	} else if (!info.force_bindery_login && NWIsDSServer(conn, NULL))
	{
		if ((err = nds_login_auth(conn, spec.user, spec.password)))
		{
			if (err != NWE_PASSWORD_EXPIRED) {
				mycom_err(err, _("in nds login"));
				fprintf(stderr, _("Login denied.\n"));
				ncp_close(conn);
				proc_ncpm_umount(mount_point);
				free(mount_name);
				NWDSFreeContext(ctx);
				if (auth_conn) {
					NWCCCloseConn(auth_conn);
				}
				exit(55);
			}
	    		fprintf(stderr, _("Your password has expired\n"));
		}
	}
	else
	{
#endif	
	if ((err = ncp_login_user(conn, spec.user, spec.password)) != 0)
	{
		struct nw_property p;
		struct ncp_prop_login_control *l
		= (struct ncp_prop_login_control *) &p;

		if (err != NWE_PASSWORD_EXPIRED)
		{
			mycom_err(err, _("in login"));
			fprintf(stderr, _("Login denied\n"));
			ncp_close(conn);
			proc_ncpm_umount(mount_point);
			free(mount_name);
			NWDSFreeContext(ctx);
			if (auth_conn) {
				NWCCCloseConn(auth_conn);
			}
			exit(56);
		}
		fprintf(stderr, _("Your password has expired\n"));

		if ((err = ncp_read_property_value(conn, NCP_BINDERY_USER,
						   spec.user, 1,
						   "LOGIN_CONTROL", &p)) == 0)
		{
			fprintf(stderr, _("You have %d login attempts left\n"),
				l->GraceLogins);
		}
	}
#ifdef NDS_SUPPORT
	}
#endif
	err = NWDSFreeContext(ctx);
	if (err) {
		ncp_close(conn);
		proc_ncpm_umount(mount_point);
		free(mount_name);
		if (auth_conn) {
			NWCCCloseConn(auth_conn);
		}
		mycom_err(err, _("in free context"));
		exit(116);
	}
	if (auth_conn) {
		NWCCCloseConn(auth_conn);
	}
	if ((err = ncp_mount_specific(conn, -1, info.NWpath, info.pathlen)) != 0) 
	{
		ncp_close(conn);
		proc_ncpm_umount(mount_point);
		free(mount_name);
		errexit(57, _("Cannot access path \"%s\": %s\n"), info.remote_path, strerror(-err));
	}
	ncp_close(conn);

	if (!opt_n) {
		block_sigs();
		add_mnt_entry(mount_name, mount_point, info.flags);
		unblock_sigs();
	}
	return 0;
}

static void usage(void)
{
	printf(_("usage: %s [options] mount-point\n"), progname);
	printf(_("Try `%s -h' for more information\n"), progname);
}

static void
help(void)
{
	printf(_("\n"
		 "usage: %s [options] mount-point\n"), progname);
	/* printf() is macro in glibc2.2... */
	fprintf(stdout,
	     _("\n"
	       "-S server      Server name to be used\n"
	       "-A dns_name    DNS server name to be used when mounting over TCP or UDP\n"
	       "-U username    Username sent to server\n"
	       "-V volume      Volume to mount, for NFS re-export\n"
	       "-u uid         uid the mounted files get\n"
	       "-g gid         gid the mounted files get\n"
	       "-f mode        permission the files get (octal notation)\n"
	       "-d mode        permission the dirs get (octal notation)\n"
	       "-c uid         uid to identify the connection to mount on\n"
	       "               Only makes sense for root\n"
	       "-t time_out    Waiting time (in 1/100s) to wait for\n"
	       "               an answer from the server. Default: 60\n"
	       "-r retry_count Number of retry attempts. Default: 5\n"
	       "-C             Don't convert password to uppercase\n"
	       "-P password    Use this password\n"
	       "-n             Do not use any password\n"
	       "               If neither -P nor -n are given, you are\n"
	       "               asked for a password.\n"
	       "-s             Enable renaming/deletion of read-only files\n"
	       "-h             print this help text\n"
	       "-v             print ncpfs version number\n"
	       "%s%s"
	       "-m             Allow multiple logins to server\n"
	       "-N os2,nfs     Do not use specified namespaces on mounted volume\n"
	       "-y charset     character set used for input and display\n"
	       "-p codepage    codepage used on volume, including letters `cp'\n"
	       "\n"),
#ifdef NDS_SUPPORT
	       _("-b             Force bindery login to NDS servers\n")
#else
	       ""
#endif /* NDS_SUPPORT */
	       ,
#ifdef SIGNATURES
	       _("-i level       Signature level, 0=never, 1=supported, 2=preferred, 3=required\n")
#else
	       ""
#endif /* SIGNATURES */
	       );
}
